include/blmc_robots/real_finger.hpp

```diff
@@ -373,4 +402,253 @@ robot_interfaces::finger::BackendPtr create_real_finger_backend(
+        observation.velocity[1] = 2 * i + 1;
+        observation.velocity[2] = 3 * i + 1;
+        observation.torque[0] = i + 2;
+        observation.torque[1] = 2 * i + 2;
+        observation.torque[2] = 3 * i + 2;
+
+        i++;
+
+        return observation;
+    }
+
+    // check for better "void" implementations
+
+    Action apply_action(const Action &desired_action) override
+    {
+        // (maybe) isolated casue to this: the random driver just goes on
+        // spitting values,
+        //  and by the time one observation is written to the file, 10 newer
+        //  ones get
+        // added to the buffer already. View this by printing out the index and
+        // the logger_data_->observation->newest_timeindex_()
+        double start_time_sec = real_time_tools::Timer::get_current_time_sec();
+
+        real_time_tools::Timer::sleep_until_sec(start_time_sec + 0.0001);
+
+        // okay, so reducing the delay here actually helped. Seems
+        // counterintuitive rn.
+
+        return desired_action;
+    }
+
+    void shutdown() override
+    {
+        return;
+    }
+};
+
+robot_interfaces::finger::BackendPtr create_random_finger_backend(
+    robot_interfaces::finger::DataPtr robot_data)
+{
+    std::shared_ptr<
+        robot_interfaces::RobotDriver<robot_interfaces::finger::Action,
+                                      robot_interfaces::finger::Observation>>
+        robot = std::make_shared<RandomFingerDriver>();
+
+    auto backend =
+        std::make_shared<robot_interfaces::finger::Backend>(robot, robot_data);
+    backend->set_max_action_repetitions(-1);
+
+    return backend;
+}
+
+class Logger
+{
+    // template <typename Type>
+public:
+    int block_size_ = 10;
+    int count = 0;
+
+    typedef robot_interfaces::finger::Action Action;
+    typedef robot_interfaces::finger::Observation Observation;
+    typedef robot_interfaces::finger::Status Status;
+    // typedef std::numeric_limits< double > dbl;
+
+    bool destructor_was_called_;
+
+    // The Logger
+    Logger(std::shared_ptr<
+           robot_interfaces::RobotData<Action, Observation, Status>> robot_data)
+        : logger_data_(robot_data), destructor_was_called_(false)
+    {
+        thread_ = std::make_shared<real_time_tools::RealTimeThread>();
+    }
+
+    std::shared_ptr<robot_interfaces::RobotData<Action, Observation, Status>>
+        logger_data_;
+
+    virtual ~Logger()
+    {
+        destructor_was_called_ = true;
+        thread_->join();
+    }
+
+    int index;
+
+    static void *write(void *instance_pointer)
+    {
+        ((Logger *)(instance_pointer))->write();
+        return nullptr;
+    }
+
+    void write()
+    {
+        // std::ofstream
+        // output_file("/local_data/sjoshi/blmc_ei/workspace/src/catkin/robots/blmc_robots/demos/./log.csv");
+        std::ofstream output_file;
+        // std::cout << "Hello";
+
+        // errno = 0;
+        output_file.open("log.csv");
```

Comment:
The filename should be configurable.  I think the best would be to pass it as an argument to the constructor.